backend/src/main/java/com/huellapositiva/domain/actions/FetchCredentialsAction.java

```diff
@@ -0,0 +1,83 @@
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.Calendar;
+import java.util.Date;
+
+
+@Service
+@AllArgsConstructor
+public class FetchCredentialsAction {
+
+    @Autowired
+    EmailCommunicationService emailCommunicationService;
+
+    @Autowired
+    private JpaCredentialRepository jpaCredentialRepository;
+
+    @Autowired
+    private final PasswordEncoder passwordEncoder;
+
+    /**
+     * This method generates a link to recovery the password and sends it
+     *
+     * @param email Email account from de volunteer or contact person.
+     */
+    public void executeGenerationRecoveryPasswordEmail(String email) {
+        JpaCredential jpaCredential = jpaCredentialRepository.findByEmail(email).orElseThrow(UserNotFoundException::new);
+        String hash = Token.createToken().toString();
+        jpaCredentialRepository.updateHashByEmail(email, hash);
+        EmailRecoveryPassword emailRecoveryPassword = EmailRecoveryPassword.from(jpaCredential.getEmail(), hash);
+        emailCommunicationService.sendRecoveryPasswordEmail(emailRecoveryPassword);
+    }
+
+    /**
+     * This method updates the password, updates the hashRecoveryPassword and createdRecoveryHashOn to null in database
+     * and sends a confirmation email.
+     *
+     * @param hashRecoveryPassword Unique identifier to recover password
+     * @param password New password
+     */
+    public void executePasswordChanging(String hashRecoveryPassword, String password) {
+        JpaCredential jpaCredential = jpaCredentialRepository.findByHashRecoveryPassword(hashRecoveryPassword).orElseThrow(UserNotFoundException::new);
+        Date timeOfExpiration = addAnHour(jpaCredential.getCreatedRecoveryHashOn());
+        Date dateNow = Calendar.getInstance().getTime();
+
+        if (timeOfExpiration.after(dateNow)) {
+            PasswordHash passwordHash = new PasswordHash(passwordEncoder.encode(password));
+            jpaCredentialRepository.updatePassword(passwordHash.toString(), jpaCredential.getEmail());
+            jpaCredentialRepository.setRecoveryPasswordHashAndDate(jpaCredential.getEmail(), null, null);
+
+            EmailAddress emailAddress = EmailAddress.from(jpaCredential.getEmail());
+            emailCommunicationService.sendConfirmationPasswordChanged(emailAddress);
+        } else {
+            throw new TimeForRecoveringPasswordExpiredException("The time to recovery password has expired");
+        }
+    }
+
+    /**
+     * This method adds an hour to the date
+     *
+     * @param date
+     * @return It is the date with an hour added.
+     */
+    private Date addAnHour(Date date) {
+        Calendar c = Calendar.getInstance();
+        c.setTime(date);
+        c.add(Calendar.HOUR, 1);  // number of days to add
+        return c.getTime();
+    }
```

Comment:
Java Time API has built-in functionality for this. Don't reinvent the wheel :-)

Date and Calendar objects are obsolete, favour using the new Time API -> Instant, LocalDate, LocalDateTime, ...

Reply:
Ok Ibai, we have updated the Date/Calendar in Database to LocalDateTime. Many thanks, please review it.